crates/rethnet_rpc_server/src/lib.rs

```diff
@@ -64,13 +64,63 @@ pub enum MethodInvocation {
 
 type RethnetStateType = Arc<RwLock<dyn SyncState<StateError>>>;
 
+struct _FilterCriteria {
+    _from_block: U256,
+    _to_block: U256,
+    _addresses: Vec<Address>,
+    _topics: Vec<B256>,
+}
+
+impl _FilterCriteria {
+    async fn _from_request_and_state<T>(
+        request_options: FilterOptions,
+        state: StateType,
+    ) -> Result<Self, ResponseData<T>> {
```

Comment:
Why does this use a `ResponseData<T>` as error type? Shouldn't this just be a json-rpc error in that case?

Reply:
It's because a json-rpc error (that is, an error that's returned to a json-rpc client) _is_ a `ResponseData<T>`, or, more specifically, a variant of it: `ResponseData::<T>::Error`, as returned by [`ResponseData<T>::new_error()`](https://github.com/NomicFoundation/hardhat/blob/c04ff55b42d7d8c23b6f1f57757b9a2b444688c4/crates/rethnet_eth/src/remote/jsonrpc.rs#L62), which is used by the `error_response_data` function that's local to this file, which in turn will be used by the `_block_number_from_hash` and `_block_number_from_block_spec` functions that this method uses, when those functions are implemented.

Reply:
In any case, I'd consider this subject to change when I go to implement the "complex filters", and I'd prefer to defer any further refinement until we've got some code that's actually using this function.

Reply:
That's fair. Sounds good to me